p2p/discovery/server.go

```diff
@@ -8,59 +8,67 @@ import (
 	"github.com/seeleteam/go-seele/common"
 	"net"
 	"sync"
+	"time"
 
-	"github.com/seeleteam/go-seele/crypto"
 	"github.com/seeleteam/go-seele/log"
 )
 
-func StartServer(port string) {
-	udp := getUDP(port)
-	log.Debug("nodeid:" + common.BytesToHex(udp.self.ID.Bytes()))
+const (
+	PINGPONGINTERVER = 10 * time.Second // sleep between ping pong
+	DISCOVERYINTERVER = 10 * time.Second // sleep between discovery
+)
+
+func StartServer(port, id string) {
+	udp := getUDP(port, HexToNodeID(id))
+	log.Debug("nodeid: %s", common.BytesToHex(udp.self.ID.Bytes()))
 
-	udp.readLoop()
+	udp.StartServe()
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	wg.Wait()
 }
 
-func getUDP(port string) *udp {
+func getUDP(port string, id NodeID) *udp {
 	addr, err := net.ResolveUDPAddr("udp", ":"+port)
 	if err != nil {
-		log.Fatal(err)
+		log.Fatal(err.Error())
 	}
 
-	keypair, err := crypto.GenerateKey()
+	return NewUDP(id, addr)
+}
+
+func HexToNodeID(id string) NodeID  {
+	byte, err := common.HexToBytes(id)
 	if err != nil {
-		log.Info(err)
+		log.Fatal(err.Error())
 	}
 
-	buff := crypto.FromECDSAPub(&keypair.PublicKey)
-
-	id, err := BytesTOID(buff[1:])
+	nid, err := BytesTOID(byte)
 	if err != nil {
-		log.Fatal(err)
+		log.Fatal(err.Error())
 	}
 
-	return NewUDP(id, addr)
+	return nid
 }
 
-func SendPing(port string, target string) {
-	udp := getUDP(port)
+func SendPing(port, id, targePort string) {
+	myid := getRandomNodeID()
 
-	go udp.readLoop()
+	udp := getUDP(port, myid)
 
-	log.Debug("nodeid: " + common.BytesToHex(udp.self.ID.Bytes()))
+	log.Debug("nodeid: %s", common.BytesToHex(udp.self.ID.Bytes()))
 
-	addr, err := net.ResolveUDPAddr("udp", ":"+target)
-	if err != nil {
-		log.Fatal(err)
-	}
+	addr := getAddr(targePort)
+	tid := HexToNodeID(id)
 
-	msg := &Ping{
-		ID: udp.self.ID,
-	}
+	n := NewNodeWithAddr(tid, addr)
+	udp.table.addNode(n)
+	udp.db.add(n.sha, n)
 
-	udp.sendPingMsg(msg, addr)
+	udp.StartServe()
 
 	var wg sync.WaitGroup
```

Comment:
what's difference between 
var wg sync.WaitGroup 
and 
wg := sync.WaitGroup{}

Reply:
no difference

Reply:
sync.WaitGroup is using for testing